Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp

```diff
@@ -28,26 +28,144 @@
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> IntSize {
+        return { static_cast<int>(offscreenCanvas->width()), static_cast<int>(offscreenCanvas->height()) };
+    }
+#endif
+    );
+}
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(GPUCanvasContext);
+
+std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
+{
+    return std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
+}
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+    , m_currentConfiguration()
+    , m_surface()
+{
+    suspendIfNeeded();
+}
+
+void GPUCanvasContext::reshape(int width, int height)
+{
+    if (!m_currentConfiguration)
+        return;
+
+#if PLATFORM(COCOA)
+    m_surface = WebGPUSurfaceCocoa::create(width, height, m_currentConfiguration->convertToBacking());
+    ASSERT(m_surface);
+#else
+    UNUSED_PARAM(width);
+    UNUSED_PARAM(height);
+#endif
+}
+
 GPUCanvasContext::CanvasType GPUCanvasContext::canvas()
 {
-    return { };
+    return htmlCanvas();
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration&& configuration)
 {
+    m_currentConfiguration = WTFMove(configuration);
+
+    auto canvasSize = getCanvasSizeAsIntSize(htmlCanvas());
+    reshape(canvasSize.width(), canvasSize.height());
+}
 
+RefPtr<GPUTexture> GPUCanvasContext::getCurrentTexture()
+{
+    markContextChangedAndNotifyCanvasObservers();
+    return GPUTexture::create(m_surface->currentTexture());
 }
 
 void GPUCanvasContext::unconfigure()
 {
+    m_currentConfiguration.reset();
+    m_surface = nullptr;
 }
 
-RefPtr<GPUTexture> GPUCanvasContext::getCurrentTexture()
+PixelFormat GPUCanvasContext::pixelFormat() const
 {
-    return nullptr;
+    if (!m_surface)
+        return PixelFormat::BGRA8;
+
+    return m_surface->pixelFormat();
+}
+
+DestinationColorSpace GPUCanvasContext::colorSpace() const
+{
+    if (!m_surface)
+        return DestinationColorSpace::SRGB();
+
+    return m_surface->colorSpace();
+}
+
+RefPtr<GraphicsLayerContentsDisplayDelegate> GPUCanvasContext::layerContentsDisplayDelegate()
+{
+    if (!m_surface)
+        return nullptr;
+
+    return m_surface;
+}
+
+void GPUCanvasContext::prepareForDisplay()
+{
+    if (!m_surface)
+        return;
+
+    m_surface->present();
+    m_compositingResultsNeedsUpdating = false;
+}
+
+void GPUCanvasContext::markContextChangedAndNotifyCanvasObservers()
+{
+    m_compositingResultsNeedsUpdating = true;
+
+    bool canvasIsDirty = false;
+
+    if (auto* canvas = htmlCanvas()) {
+        RenderBox* renderBox = canvas->renderBox();
+        if (isAccelerated() && renderBox && renderBox->hasAcceleratedCompositing()) {
+            canvasIsDirty = true;
+            canvas->clearCopiedImage();
+            // renderBox->contentChanged(CanvasPixelsChanged);
+            renderBox->contentChanged(CanvasChanged);
+            // return;
```

Comment:
Don’t leave commented code

Reply:
I'm unsure if this is correct so for the time being I'll leave it there and wait for someone to chime in.